### ISSUE TITLE
Usage of `padding_free`

### ISSUE THREAD

**Issue by zwhe99** [huggingface/trl#2242]
I found a `padding_free` option in [`DataCollatorForCompletionOnlyLM`](https://github.com/huggingface/trl/blob/02f4e750c07c5a470f2d82a3a59e011401b5c63e/trl/trainer/utils.py#L227-L233), which concats all samples in a batch and replace padding and attention masking with position ids.

It seems useful for accelerating training, but I found it less discussed in the doc.

Is this an experimental feature and can I use this safely?

@RhuiDih @lewtun @kashif

**Comment by qgallouedec** [comment]
You can use it, feel free to report if it causes any issues.

**Comment by zwhe99** [comment]
Thanks for the response!

**Comment by shiym2000** [comment]
@qgallouedec 
When using `padding_free`, do the parameters `per_device_train_batch_size` and `max_length` still take effect? If so, how exactly are they applied in processing a batch of data?

Furthermore, if a custom `data_collator` is passed into the trainer, does it mean that `padding_free` cannot be utilized unless the functionality of `padding_free` is explicitly implemented within the custom `data_collator`?

**Comment by qgallouedec** [comment]
> When using padding_free, do the parameters per_device_train_batch_size and max_length still take effect?

yes, check the doc https://huggingface.co/docs/trl/en/reducing_memory_usage#padding-free

> Furthermore, if a custom data_collator is passed into the trainer, does it mean that padding_free cannot be utilized unless the functionality of padding_free is explicitly implemented within the custom data_collator?

Yes exactly

**Comment by shiym2000** [comment]
@qgallouedec Thank you for your reply! After reviewing the doc, I believe that in the **padding_free** method, `per_device_train_batch_size` is required, but the role of `max_length` is not reflected in the doc. I was wondering—if we truncate sequences according to `max_length`, wouldn’t this essentially become the same as the **packing** method? I’m not sure which parts of my understanding are correct and which are mistaken.

In addition, if **padding_free** needs to be implemented manually in a custom `data_collator`, does that also mean that both **truncation** and **packing** must be implemented manually within the custom collator?

**Comment by qgallouedec** [comment]
> but the role of max_length is not reflected in the doc

the max_length is applied before flattening the data. so if you have `max_length=4` and `per_device_batch_size=2` then this batch

```
X X X X X X
Y Y Y
```

becomes


```
X X X X Y Y Y
```

> truncate sequences according to max_length, wouldn’t this essentially become the same as the packing method?

No. When `padding_free` is enabled the sampled data is flattened (instead of padded). So if you count the number of different samples that you have in your final batch (which is flattened), it's always `per_device_batch_size`. In the above example, you've 2 samples in the final batch.

When `packing` is enabled, the dataset is prepared so that if two samples can fit into `max_lenght`, they are gathered into one. Basically, the dataset turn from (`max_length=6`):

```
W W
X X X X X X
Y Y Y
Z Z Z Z Z
```

into

```
W W Y Y Y
X X X X X X
Z Z Z Z Z
```

and then, when sampled, with eg `per_device_batch_size=2`:

```
W W Y Y Y
X X X X X X
```

padding-free is applied (`packing=True` implies `padding_free=True):

```
W W Y Y Y X X X X X X
```

You end up with a sample that can contain an arbitrary number of different samples.





> but the role of `max_length` is not clear in the doc

The `max_length` parameter is applied before flattening the data. For example, if you have `max_length=4` and `per_device_batch_size=2` and your batch looks like this:

```
X X X X X X
Y Y Y
```

It will be flattened to:

```
X X X X Y Y Y
```

> truncate sequences according to max_length, wouldn’t this essentially become the same as the packing method?

Not exactly. Here’s the difference:

1. Padding-free:
   * The sampled data is flattened instead of padded.
   * The final batch always contains exactly `per_device_batch_size` distinct samples.
   * Using the previous example, the final batch still has 2 samples (X and Y) after flattening.

2. Packing:
The dataset is preprocessed so that multiple samples can be combined if they fit within `max_length`. Example dataset with `max_length=6`:

 ```
 W W
 X X X X X X
 Y Y Y
 Z Z Z Z Z
 ```

After packing, the dataset becomes:

 ```
 W W Y Y Y
 X X X X X X
 Z Z Z Z Z
 ```

Sampling with `per_device_batch_size=2`:

 ```
 W W Y Y Y
 X X X X X X
 ```

`padding_free` is applied (`packing=True` automatically enables `padding_free`):

 ```
 W W Y Y Y X X X X X X
 ```

The final batch can now contain an arbitrary number of original samples, depending on how they fit into `max_length`.

**Comment by shiym2000** [comment]
@qgallouedec Thank you very much for your insightful reply!

I have one more question: If a custom data_collator is passed to the trainer, such as SFTTrainer, do I need to manually implement the three methods (truncation, packing, padding-free) for reducing memory usage in the custom data_collator? 

If so, could you please point me to any relevant code references?

**Comment by qgallouedec** [comment]
So, the data collator gets the sampled data, you only need to implement padding-free

**Comment by shiym2000** [comment]
@qgallouedec Thank you for your reply! However, during SFT with Qwen2.5-VL, when I passed in a custom `data_collator`, changing the `max_length` parameter did not trigger the expected truncation (without using packing). Likewise, when I passed in the `packing` parameter, it also had no effect.

Here is the relevant code:
``` python
def collate_func(examples: list[dict[str, Any]], processor, mode) -> dict[str, torch.Tensor]:
    texts = []
    images = []
    videos = []

    for example in examples:
        texts.append(processor.apply_chat_template(
            example["messages"],
            tokenize=False,
            add_generation_prompt=True if mode == "eval" else False,
        ))

        image_inputs, video_inputs = process_vision_info(example["messages"])
        if image_inputs is not None:
            images.extend(image_inputs)
        if video_inputs is not None:
            videos.extend(video_inputs)  # list: [T, 3, H, W], [0, 255]

    if len(images) == 0:
        images = None
    if len(videos) == 0:
        videos = None

    if mode == "eval_vllm":
        return {
            "texts": texts,
            "images": images,
            "videos": videos,
        }

    batch = processor(
        text=texts,
        images=images,
        videos=videos,
        padding=True,
        return_tensors="pt",
    )  # input_ids, attention_mask, pixel_values_videos, video_grid_thw, second_per_grid_ts

    if mode == "eval":
        return batch

    # Choice 1
    if mode == "pretrain":
        labels = batch["input_ids"].clone()

    # Choice 2: supervise assistant response (same with ms-swift)
    # <|im_start|>assistant\n...<|im_end|>\n
    # [151644, 77091, 198] ... [151645, 198]
    elif mode == "sft":
        labels = torch.full_like(batch["input_ids"], -100)
        B, L = batch["input_ids"].shape
        for input_ids_cur, labels_cur in zip(batch["input_ids"], labels):
            start_idx = 0
            end_idx = 0
            while start_idx < L:
                if input_ids_cur[start_idx] == processor.tokenizer.encode("<|im_start|>")[0]:
                    if input_ids_cur[start_idx + 1] == processor.tokenizer.encode("assistant")[0]:
                        start_idx = start_idx + len(processor.tokenizer.encode("<|im_start|>assistant\n"))
                        end_idx = start_idx + 1
                        while input_ids_cur[end_idx] != processor.tokenizer.encode("<|im_end|>")[0]:
                            end_idx = end_idx + 1
                        labels_cur[start_idx:end_idx+1] = input_ids_cur[start_idx:end_idx+1]
                start_idx = start_idx + 1

    # pad others
    labels[labels == processor.tokenizer.pad_token_id] = -100

    # Handle visual tokens based on processor type
    # <|vision_start|> <|vision_end|> <|image_pad|> <|video_pad|>
    visual_tokens = (
        [151652, 151653, 151655, 151656]
        if isinstance(processor, Qwen2VLProcessor)
        else [processor.tokenizer.convert_tokens_to_ids(processor.image_token)]
    )
    for visual_token_id in visual_tokens:
        labels[labels == visual_token_id] = -100

    batch["labels"] = labels
    return batch

collate_fn = partial(
    collate_func,
    processor=processor,
    mode="sft",
)
trainer = SFTTrainer(
    model=model,
    args=training_args,
    train_dataset=train_dataset,
    data_collator=collate_fn,
    processing_class=processor,
    peft_config=get_peft_config(model_args),
)
```

**[Baseline]** When I set max_length=8192 (without packing):
``` bash
nproc_per_node=1
CUDA_VISIBLE_DEVICES=0 \
accelerate launch \
    --config_file sym/deepspeed_zero2.yaml \
    sym/sft_qwenvl_3d.py \
    --model_name_or_path /yinghepool/zhangshuheng/models/Qwen2.5-VL-3B-Instruct \
    --attn_implementation flash_attention_2 \
    --torch_dtype bfloat16 \
    --dataset_name /data/jiawei_workspace/report_generate_docker/exps/dataset/swift/train_rg_modified_swift.json \
    --output_dir $MODEL_PATH \
    --num_train_epochs 1 \
    --per_device_train_batch_size 4 \
    --gradient_accumulation_steps $(expr 16 / $nproc_per_node) \
    --learning_rate 2e-5 \
    --warmup_ratio 0.03 \
    --lr_scheduler_type cosine \
    --max_length 8192 \  # key parameter
    --dataloader_num_workers 8 \
    --bf16 True \
    --gradient_checkpointing True \
    --logging_strategy steps \
    --logging_steps 1 \
    --save_strategy steps \
    --save_steps 50 \
    --save_total_limit 1 \
    --report_to tensorboard
```

<img width="1966" height="182" alt="Image" src="https://github.com/user-attachments/assets/47041341-9235-4dd9-8f66-dfe46572f2ce" />


**[Exp1]** When I set max_length=2800 (without packing):
``` bash
nproc_per_node=1
CUDA_VISIBLE_DEVICES=0 \
accelerate launch \
    --config_file sym/deepspeed_zero2.yaml \
    sym/sft_qwenvl_3d.py \
    --model_name_or_path /yinghepool/zhangshuheng/models/Qwen2.5-VL-3B-Instruct \
    --attn_implementation flash_attention_2 \
    --torch_dtype bfloat16 \
    --dataset_name /data/jiawei_workspace/report_generate_docker/exps/dataset/swift/train_rg_modified_swift.json \
    --output_dir $MODEL_PATH \
    --num_train_epochs 1 \
    --per_device_train_batch_size 4 \
    --gradient_accumulation_steps $(expr 16 / $nproc_per_node) \
    --learning_rate 2e-5 \
    --warmup_ratio 0.03 \
    --lr_scheduler_type cosine \
    --max_length 2800 \  # key parameter
    --dataloader_num_workers 8 \
    --bf16 True \
    --gradient_checkpointing True \
    --logging_strategy steps \
    --logging_steps 1 \
    --save_strategy steps \
    --save_steps 50 \
    --save_total_limit 1 \
    --report_to tensorboard
```

<img width="1974" height="143" alt="Image" src="https://github.com/user-attachments/assets/b13e044e-6b73-495e-9196-8f5ab4f7d998" />


**[Exp2]** When I set max_length=8192 (with packing):
``` bash
nproc_per_node=1
CUDA_VISIBLE_DEVICES=0 \
accelerate launch \
    --config_file sym/deepspeed_zero2.yaml \
    sym/sft_qwenvl_3d.py \
    --model_name_or_path /yinghepool/zhangshuheng/models/Qwen2.5-VL-3B-Instruct \
    --attn_implementation flash_attention_2 \
    --torch_dtype bfloat16 \
    --dataset_name /data/jiawei_workspace/report_generate_docker/exps/dataset/swift/train_rg_modified_swift.json \
    --output_dir $MODEL_PATH \
    --num_train_epochs 1 \
    --per_device_train_batch_size 4 \
    --gradient_accumulation_steps $(expr 16 / $nproc_per_node) \
    --learning_rate 2e-5 \
    --warmup_ratio 0.03 \
    --lr_scheduler_type cosine \
    --max_length 8192 \  # key parameter
    --packing True \  # key parameter
    --packing_strategy wrapped \  # key parameter
    --padding_free False \  # key parameter
    --dataloader_num_workers 8 \
    --bf16 True \
    --gradient_checkpointing True \
    --logging_strategy steps \
    --logging_steps 1 \
    --save_strategy steps \
    --save_steps 50 \
    --save_total_limit 1 \
    --report_to tensorboard
```

<img width="1971" height="180" alt="Image" src="https://github.com/user-attachments/assets/1fda2d5f-d624-4cfc-a2f7-3bed8607902c" />

**Comment by adosar** [comment]
> > but the role of max_length is not reflected in the doc
> 
> the max_length is applied before flattening the data. so if you have `max_length=4` and `per_device_batch_size=2` then this batch
> 
> ```
> X X X X X X
> Y Y Y
> ```
> 
> becomes
> 
> ```
> X X X X Y Y Y
> ```
> 
> > truncate sequences according to max_length, wouldn’t this essentially become the same as the packing method?
> 
> No. When `padding_free` is enabled the sampled data is flattened (instead of padded). So if you count the number of different samples that you have in your final batch (which is flattened), it's always `per_device_batch_size`. In the above example, you've 2 samples in the final batch.
> 
> When `packing` is enabled, the dataset is prepared so that if two samples can fit into `max_lenght`, they are gathered into one. Basically, the dataset turn from (`max_length=6`):
> 
> ```
> W W
> X X X X X X
> Y Y Y
> Z Z Z Z Z
> ```
> 
> into
> 
> ```
> W W Y Y Y
> X X X X X X
> Z Z Z Z Z
> ```
> 
> and then, when sampled, with eg `per_device_batch_size=2`:
> 
> ```
> W W Y Y Y
> X X X X X X
> ```
> 
> padding-free is applied (`packing=True` implies `padding_free=True):
> 
> ```
> W W Y Y Y X X X X X X
> ```
> 
> You end up with a sample that can contain an arbitrary number of different samples.
> 
> > but the role of `max_length` is not clear in the doc
> 
> The `max_length` parameter is applied before flattening the data. For example, if you have `max_length=4` and `per_device_batch_size=2` and your batch looks like this:
> 
> ```
> X X X X X X
> Y Y Y
> ```
> 
> It will be flattened to:
> 
> ```
> X X X X Y Y Y
> ```
> 
> > truncate sequences according to max_length, wouldn’t this essentially become the same as the packing method?
> 
> Not exactly. Here’s the difference:
> 
>     1. Padding-free:
>        
>        * The sampled data is flattened instead of padded.
>        * The final batch always contains exactly `per_device_batch_size` distinct samples.
>        * Using the previous example, the final batch still has 2 samples (X and Y) after flattening.
> 
>     2. Packing:
>        The dataset is preprocessed so that multiple samples can be combined if they fit within `max_length`. Example dataset with `max_length=6`:
> 
> 
> ```
> W W
> X X X X X X
> Y Y Y
> Z Z Z Z Z
> ```
> 
> After packing, the dataset becomes:
> 
> ```
> W W Y Y Y
> X X X X X X
> Z Z Z Z Z
> ```
> 
> Sampling with `per_device_batch_size=2`:
> 
> ```
> W W Y Y Y
> X X X X X X
> ```
> 
> `padding_free` is applied (`packing=True` automatically enables `padding_free`):
> 
> ```
> W W Y Y Y X X X X X X
> ```
> 
> The final batch can now contain an arbitrary number of original samples, depending on how they fit into `max_length`.

Hi @qgallouedec ,

If `packing=True` enables `padding_free` does this mean that one of them is redundant? I think my confusion arises from the following image in the [docs](https://huggingface.co/docs/trl/reducing_memory_usage?liger=SFT#packing):

<img width="500" height="600" alt="Image" src="https://github.com/user-attachments/assets/582dd8e4-0f7d-4504-8f0a-01b7e9cd218a" />

As you can see in the right image (`packing=True`), one sequence can be split into one or more rows or even be in the same row with another sequence. 

1. Does the attention mask account's for that? E.g. will the tokens of the second row attend to the tokens of the first row?
2. There are still padding tokens so no `padding_free`?
3. When fine-tuning a pretrained model do we need to know if it was trained with padding or packing?

**Comment by qgallouedec** [comment]
> Does the attention mask account's for that?

yes, but it requires flash attention

> There are still padding tokens so no padding_free?

let's say almost 0 padding

> When fine-tuning a pretrained model do we need to know if it was trained with padding or packing?

no, it shouldn't change anything
